### PR TITLE
structural and content changes to the sql_database doc

### PR DESCRIPTION
**Pull request by rahuljo**

### CODE REVIEW

**Review by VioletM — DISMISSED** [review]

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md [hunk 1]
```diff
@@ -282,109 +420,95 @@ info = pipeline.run(
...
-           initial_value=datetime(2024, 1, 1)  # Initial cursor value
+           initial_value=pendulum.DateTime(2024, 1, 1, 0, 0, 0)  # Initial cursor value
        )
    )
 
    info = pipeline.extract(table, write_disposition="merge")
    print(info)
    ```
+    Behind the scene, the loader generates a SQL query filtering rows with `last_modified` values greater than the incremental value. In the first run, this is the initial value (midnight (00:00:00) January 1, 2024). 
+    In subsequent runs, it is the latest value of "last_modified" that `dlt` stores in [state](https://dlthub.com/docs/general-usage/state).
```
Comment (VioletM) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md:484]:
```suggestion
    In subsequent runs, it is the latest value of `last_modified` that `dlt` stores in [state](https://dlthub.com/docs/general-usage/state).
```

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md [hunk 2]
```diff
@@ -282,109 +420,95 @@ info = pipeline.run(
...
 ```py
 database = sql_database().parallelize()
 table = sql_table().parallelize()
 ```
 
-## Column reflection
-
+### Column reflection
 Columns and their data types are reflected with SQLAlchemy. The SQL types are then mapped to `dlt` types.
 Most types are supported.
```
Comment (VioletM) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md:513]:
```suggestion
Depending on the selected backend, some of the types might require additional processing.
```

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md [hunk 3]
```diff
@@ -394,29 +518,24 @@ The `reflection_level` argument controls how much information is reflected:
 - `reflection_level = "full"`: Column names, nullability, and data types are detected. For decimal types we always add precision and scale. **This is the default.**
 - `reflection_level = "full_with_precision"`: Column names, nullability, data types, and precision/scale are detected, also for types like text and binary. Integer sizes are set to bigint and to int for all other types.
```
Comment (VioletM) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md:519]: Not sure I understand the latest sentence, could you elaborate?

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md [hunk 4]
```diff
@@ -394,29 +518,24 @@ The `reflection_level` argument controls how much information is reflected:
...
 
-If the SQL type is unknown or not supported by `dlt` the column is skipped when using the `pyarrow` backend.
-In other backend the type is inferred from data regardless of `reflection_level`, this often works, some types are coerced to strings
-and `dataclass` based values from sqlalchemy are inferred as `complex` (JSON in most destinations).
-
+If the SQL type is unknown or not supported by `dlt`, then, in the pyarrow backend, the column will be skipped, whereas in the other backends the type will be inferred directly from the data irrespective of the `reflection_level` specified. In the latter case, this often means that some types are coerced to strings and  `dataclass` based values from sqlalchemy are inferred as `complex` (JSON in most destinations).
```
Comment (VioletM) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md:521]: I'm not sure this is true for pyarrow anymore, could you please double check?

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md [hunk 5]
```diff
@@ -394,29 +518,24 @@ The `reflection_level` argument controls how much information is reflected:
...
 
-This is useful for example when:
+This is useful, for example, when:
 - You're loading a data type which is not supported by the destination (e.g. you need JSON type columns to be coerced to string)
-- You're using an sqlalchemy dialect which uses custom types that don't inherit from standard sqlalchemy types.
+- You're using a sqlalchemy dialect which uses custom types that don't inherit from standard sqlalchemy types.
 - For certain types you prefer `dlt` to infer data type from the data and you return `None`
 
-Example, when loading timestamps from Snowflake you can make sure they translate to `timestamp` columns in the result schema:
+Example, when loading timestamps from Snowflake, you ensure that they get translated into standard sqlalchemy `timestamp` columns in the resultant schema:
```
Comment (VioletM) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md:538]:
```suggestion
In the following example, when loading timestamps from Snowflake, you ensure that they get translated into standard sqlalchemy `timestamp` columns in the resultant schema:
```

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md [hunk 6]
```diff
@@ -438,211 +557,120 @@ source = sql_database(
...
+    cursor_path="updated_at"
+    ```    
+
+    The resulting source created below will extract data using **pandas** backend with **chunk_size** 1000. The table **chat_message** will load data incrementally using **updated_at** column. All the other tables will not use incremental loading, and will instead load the full data.
+
+    ```py
+    database = sql_database()
+    ```
+
+You'll be able to configure all the arguments this way (except adapter callback function). [Standard dlt rules apply](https://dlthub.com/docs/general-usage/credentials/configuration#configure-dlt-sources-and-resources).
```
Comment (VioletM) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md:601]: broken link

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md [hunk 7]
```diff
@@ -438,211 +557,120 @@ source = sql_database(
...
+
+    The resulting source created below will extract data using **pandas** backend with **chunk_size** 1000. The table **chat_message** will load data incrementally using **updated_at** column. All the other tables will not use incremental loading, and will instead load the full data.
+
+    ```py
+    database = sql_database()
+    ```
+
+You'll be able to configure all the arguments this way (except adapter callback function). [Standard dlt rules apply](https://dlthub.com/docs/general-usage/credentials/configuration#configure-dlt-sources-and-resources).  
+  
+It is also possible to set these arguments as environment variables [using the proper naming convention](https://dlthub.com/docs/general-usage/credentials/config_providers#toml-vs-environment-variables):
```
Comment (VioletM) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md:603]: broken link

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md [hunk 8]
```diff
@@ -438,211 +557,120 @@ source = sql_database(
...
-   source = sql_database().with_resources("family")
-   # modify this source instance's resource
-   source = source.family.add_map(remove_columns)
-   # Run the pipeline. For a large db this may take a while
-   info = pipeline.run(source, write_disposition="replace")
-   print(info)
-   ```
+### Running on Airflow
+When running on Airflow:
+1. Use the `dlt` [Airflow Helper](../../walkthroughs/deploy-a-pipeline/deploy-with-airflow-composer.md#2-modify-dag-file) to create tasks from the `sql_database` source. You should be able to run table extraction in parallel with `parallel-isolated` source->DAG conversion.
```
Comment (VioletM) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md:615]: I think there is not enough context for `parallel-isolated` source->DAG conversion. Do we have somewhere explanation what does that mean? If we do, could you add a specific link here. If we don't I think we either should ask Aman to create an example, or, if it's easy then have an example here

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md [hunk 9]
```diff
@@ -438,211 +557,120 @@ source = sql_database(
...
+2. Reflect tables at runtime with `defer_table_reflect` argument.
+3. Set `allow_external_schedulers` to load data using [Airflow intervals](../../general-usage/incremental-loading.md#using-airflow-schedule-for-backfill-and-incremental-loading).
 
-1. Remember to keep the pipeline name and destination dataset name consistent. The pipeline name is crucial for retrieving the [state](https://dlthub.com/docs/general-usage/state) from the last run, which is essential for incremental loading. Altering these names could initiate a "[full_refresh](https://dlthub.com/docs/general-usage/pipeline#do-experiments-with-full-refresh)", interfering with the metadata tracking necessary for [incremental loads](https://dlthub.com/docs/general-usage/incremental-loading).
+### Transforming the data before load
+You have direct access to the extracted data through the resource objects (`sql_table()` or `sql_database().with_resource())`), each of which represents a single SQL table. These objects are generators that yield
```
Comment (VioletM) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database.md:620]: Please add a note somewhere, that with pyarrow BE, you need to adjus function you pass to `add_map` to work with pyarrow chunks. Reference: https://dlthub-community.slack.com/archives/C04DQA7JJN6/p1721686911160439

**Comment by sh-rp** [comment]
One comment from my side: we are missing the info about being able to change the sqlalchemy query in a callback for filtering, ordering and sorting.

**Comment by akelad** [comment]
@sh-rp we can add that, but tbh that should have happened from the developer that initially implemented it, right?

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database/advanced.md [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+1. **Choose a Cursor Column**: Identify a column in your SQL table that can serve as a reliable indicator of new or updated rows. Common choices include timestamp columns or auto-incrementing IDs.
+1. **Set an Initial Value**: Choose a starting value for the cursor to begin loading data. This could be a specific timestamp or ID from which you wish to start loading data.
+1. **Deduplication**: When using incremental loading, the system automatically handles the deduplication of rows based on the primary key (if available) or row hash for tables without a primary key.
+1. **Set end_value for backfill**: Set `end_value` if you want to backfill data from
+certain range.
+1. **Order returned rows**. Set `row_order` to `asc` or `desc` to order returned rows.
+
+#### Examples
+
+1. Incremental loading with the resource `sql_table`
```
Comment (mariarice15) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database/advanced.md:30]:
```suggestion
**1. Incremental loading with the resource `sql_table`**  

```

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database/advanced.md [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+       )
+   )
+
+   info = pipeline.extract(table, write_disposition="merge")
+   print(info)
+   ```
+    Behind the scene, the loader generates a SQL query filtering rows with `last_modified` values greater than the incremental value. In the first run, this is the initial value (midnight (00:00:00) January 1, 2024). 
+    In subsequent runs, it is the latest value of `last_modified` that `dlt` stores in [state](https://dlthub.com/docs/general-usage/state).
+
+2. Incremental loading with the source `sql_database`
```
Comment (mariarice15) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database/advanced.md:51]:
```suggestion
**2. Incremental loading with the source `sql_database`**  
 
```

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database/advanced.md [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+    source = sql_database().with_resources("family")
+    #using the "last_modified" field as an incremental field using initial value of midnight January 1, 2024
+    source.family.apply_hints(incremental=dlt.sources.incremental("updated"),initial_value=pendulum.DateTime(2024, 1, 1, 0, 0, 0))
+    #running the pipeline
+    info = pipeline.run(source, write_disposition="merge")
+    print(info)
+    ```  
+
+   :::info
+   * For merge write disposition, the source table needs a primary key, which `dlt` automatically sets up.
```
Comment (mariarice15) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database/advanced.md:64]:
```suggestion
   * When using "merge" write disposition, the source table needs a primary key, which `dlt` automatically sets up. 
```

File: docs/website/docs/dlt-ecosystem/verified-sources/sql_database/advanced.md [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+
+## Parallelized extraction
+
+You can extract each table in a separate thread (no multiprocessing at this point). This will decrease loading time if your queries take time to execute or your network latency/speed is low. To enable this, declare your sources/resources as follows:
+```py
+database = sql_database().parallelize()
+table = sql_table().parallelize()
+```
+
+## Column reflection
```
Comment (mariarice15) [inline docs/website/docs/dlt-ecosystem/verified-sources/sql_database/advanced.md:76]: I would add a sentence here quickly explaining reflection

**Review by mariarice15 — DISMISSED** [review]

**Review by akelad — APPROVED** [review]
approving so you can merge once it builds

**Review by akelad — APPROVED** [review]

**Pull request merged**